src/Engine/ProtoCore/DSASM/HeapObjects.cs

```diff
@@ -0,0 +1,659 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using ProtoCore.Properties;
+using ProtoCore.Runtime;
+using ProtoCore.Utils;
+
+namespace ProtoCore.DSASM
+{
+    public class DSArray : HeapElement
+    {
+        private Dictionary<StackValue, StackValue> Dict;
+        public DSArray(int size, Heap heap)
+            : base(size, heap)
+        {
+            Dict = new Dictionary<StackValue, StackValue>();
+            MetaData = new MetaData { type = (int)PrimitiveType.kTypeArray };
+        }
+
+        public IEnumerable<StackValue> Values
+        {
+            get
+            {
+                return VisibleItems.Concat(Dict.Values);
+            }
+        }
+
+        /// <summary>
+        /// Check if an array contain key
+        /// </summary>
+        /// <param name="array"></param>
+        /// <param name="key"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public bool ContainsKey(StackValue key)
+        {
+            if (key.IsNumeric)
+            {
+                long index = key.ToInteger().opdata;
+                if (index < 0)
+                {
+                    index = index + VisibleSize;
+                }
```

Comment:
I don't quite understand this, but I'm guessing it is a wrap-around thing we support in DesignScript syntax?